### PR TITLE
Rework C++ examples

### PR DESCRIPTION
**Pull request by IgorAnohin**
Hello!
It would be interesting hear your opinion about this PR.

### CODE REVIEW

File: C++/Navio/Navio+/RCInput_Navio.h [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+
+class RCInput_Navio : public RCInput
+{
+
+public:
+    RCInput_Navio();
+    void initialize();
+    int read(int ch);
+    ~RCInput_Navio();
+    void ppmOnEdge(int gpio, int level, uint32_t tick);
```
Comment (staroselskii) [inline C++/Navio/Navio+/RCInput_Navio.h]: Why it's public? Should be private. You can make this function a friend.

File: C++/Navio/Common/RCOutput.h [hunk 2]
```diff
@@ -0,0 +1,17 @@
+#ifndef RCOUTPUT
+#define RCOUTPUT
+
+#define PWN_OUTPUT 0
```
Comment (staroselskii) [inline C++/Navio/Common/RCOutput.h]: typo
s/PWN_OUTPUT/PWM_OUTPUT

And also it should be here

Comment (staroselskii) [reply C++/Navio/Common/RCOutput.h]: Please, move it where it belongs. Maybe Examples? But most definitely not library code

File: C++/Navio/Navio2/RCOutput_Navio2.cpp [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+
+RCOutput_Navio2::RCOutput_Navio2()
+{
+
+}
+
+bool RCOutput_Navio2::initialize()
+{
+    if (!pwm.init(PWN_OUTPUT)) {
+        fprintf(stderr, "Output Enable not set. Are you root?\n");
```
Comment (staroselskii) [inline C++/Navio/Navio2/RCOutput_Navio2.cpp]: Doesn't make sense. We don't have Output enable on Navio2.

File: C++/Navio/Navio2/RCOutput_Navio2.cpp [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+        return 0;
+    }
+
+    pwm.enable(PWN_OUTPUT);
+}
+
+void RCOutput_Navio2::setFrequency(float frequency)
+{
+    pwm.set_period(PWN_OUTPUT, frequency);
```
Comment (staroselskii) [inline C++/Navio/Navio2/RCOutput_Navio2.cpp]: Why only on zeroth channel? It doesn't make sense! Please.

File: C++/Navio/Common/RCOutput.h [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+
+
+class RCOutput
+{
+public:
+    virtual bool initialize() = 0;
+    virtual void setFrequency(float frequency) = 0;
```
Comment (staroselskii) [inline C++/Navio/Common/RCOutput.h]: setFrequency should work on a channel basis.

And also stick to underscores in naming.

It should be rather set_frequency.

File: C++/Examples/AccelGyroMag/AccelGyroMag.cpp [hunk 6]
```diff
@@ -47,36 +47,50 @@ void print_help()
...
+{
+    if (get_navio_version() == NAVIO2) {
+
+        if (argc < 2) {
+            printf("Enter parameter\n");
+            print_help();
+            return "-1";
+        }
+
+        // prevent the error message
```
Comment (staroselskii) [inline C++/Examples/AccelGyroMag/AccelGyroMag.cpp]: What's that?

File: C++/Examples/AccelGyroMag/AccelGyroMag.cpp [hunk 7]
```diff
@@ -47,36 +47,50 @@ void print_help()
...
+        int parameter;
+
+        while ((parameter = getopt(argc, argv, "i:h")) != -1) {
+            switch (parameter) {
+            case 'i': if (!strcmp(optarg,"mpu") ) return "mpu";
+                            else return "lsm";
+            case 'h': print_help(); return "-1";
+            case '?': printf("Wrong parameter.\n");
+                      print_help();
+                      return "-1";
```
Comment (staroselskii) [inline C++/Examples/AccelGyroMag/AccelGyroMag.cpp]: -1 is extremely bad! Please, say at least "none" or just use plain C strings and return a nullptr.

File: C++/Examples/ADC/ADC.cpp [hunk 8]
```diff
@@ -1,26 +1,55 @@
...
         return 1;
     }
 
-    while (true) {
-        for (int i = 0; i < ARRAY_SIZE(results); i++){
-            results[i] = adc.read(i);
-            printf("A%d: %.4fV ", i, results[i] / 1000);
+    if (get_navio_version() == NAVIO2) {
+
+        ADC adc{};
```
Comment (staroselskii) [inline C++/Examples/ADC/ADC.cpp]: Rename that to ADC_Navio2, add a new ADC interface and subtype. This is example code and should not deal with implementation details.

File: C++/Examples/ADC/ADC.cpp [hunk 9]
```diff
@@ -1,26 +1,55 @@
...
+                results[i] = adc.read(i);
+                printf("A%d: %.4fV ", i, results[i] / 1000);
+            }
+            printf("\n");
+
+            usleep(500000);
+        }
+    } else {
+
+        ADS1115 adc;
```
Comment (staroselskii) [inline C++/Examples/ADC/ADC.cpp]: Put all of this in ADC_Navio

File: C++/Examples/LED/LED.cpp [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+    if (check_apm()) {
+        return 1;
+    }
+
+    auto led = get_led();
+    if (led->initialize())
+        return EXIT_FAILURE;
+
+    while (true) {
+        led -> setColor(Colors::Green);
```
Comment (staroselskii) [inline C++/Examples/LED/LED.cpp]: unneeded spaces between led and setColor

File: C++/Examples/Servo/Servo.cpp [hunk 11]
```diff
@@ -14,35 +14,51 @@ sudo ./Servo
...
-    while (true) {
-        pwm.set_duty_cycle(PWM_OUTPUT, SERVO_MIN);
-        sleep(1);
-        pwm.set_duty_cycle(PWM_OUTPUT, SERVO_MAX);
-        sleep(1);
-    }
+        pwm->setFrequency(50);
+
+        while (true) {
+            pwm->set_duty_cycle(PWN_OUTPUT, SERVO_MIN); //To change the channel, you need to change PWN_OUTPUT
```
Comment (staroselskii) [inline C++/Examples/Servo/Servo.cpp]: PWN_OUTPUT

**Review by staroselskii — CHANGES_REQUESTED** [review]
Overall a big improvement. Thanks for your hard work. Please, fix these nitpickings and I'll happily merge it.

File: C++/Navio/Navio2/RCOutput_Navio2.cpp (hunk 3 above)
Comment (IgorAnohin) [reply C++/Navio/Navio2/RCOutput_Navio2.cpp]: Sorry, but I can't agree with you. Navio2 have Output enable

File: C++/Navio/Navio2/RCOutput_Navio2.cpp (hunk 4 above)
Comment (IgorAnohin) [reply C++/Navio/Navio2/RCOutput_Navio2.cpp]: Yes, you are right, I will correct all mistakes.

**Comment by IgorAnohin** [comment]
What are you think about this PR now?

File: C++/Navio/Navio2/RCOutput_Navio2.cpp [hunk 12]
```diff
@@ -0,0 +1,21 @@
+#include "RCOutput_Navio2.h"
+
+RCOutput_Navio2::RCOutput_Navio2()
+{
+}
+
+bool RCOutput_Navio2::initialize(int channel)
+{
+
+    pwm.enable(channel);
```
Comment (staroselskii) [inline C++/Navio/Navio2/RCOutput_Navio2.cpp]: missing return

File: C++/Navio/Navio2/RCOutput_Navio2.cpp [hunk 13]
```diff
@@ -0,0 +1,21 @@
...
+}
+
+void RCOutput_Navio2::set_frequency(int channel, float frequency)
+{
+    pwm.set_period(channel, frequency);
+}
+
+bool RCOutput_Navio2::set_duty_cycle(int channel, float period)
+{
+    pwm.set_duty_cycle(channel, period / 1000);
```
Comment (staroselskii) [inline C++/Navio/Navio2/RCOutput_Navio2.cpp]: ditto

File: C++/Navio/Navio2/RCOutput_Navio2.h [hunk 14]
```diff
@@ -0,0 +1,20 @@
...
+
+#include "PWM.h"
+#include <Common/RCOutput.h>
+
+
+class RCOutput_Navio2 : public RCOutput
+{
+public:
+    RCOutput_Navio2();
+    bool initialize(int channel);
```
Comment (staroselskii) [inline C++/Navio/Navio2/RCOutput_Navio2.h]: missing `override` specifier on all of the overridden methods

File: C++/Navio/Navio+/RCOutput_Navio.cpp [hunk 15]
```diff
@@ -0,0 +1,35 @@
...
+
+
+    Pin pin(outputEnablePin);
+
+    if (pin.init()) {
+        pin.setMode(Pin::GpioModeOutput);
+        pin.write(0); /* drive Output Enable low */
+    } else {
+        fprintf(stderr, "Output Enable not set. Are you root?\n");
+        return 0;
```
Comment (staroselskii) [inline C++/Navio/Navio+/RCOutput_Navio.cpp]: return false!

File: C++/Navio/Navio+/RCOutput_Navio.cpp [hunk 16]
```diff
@@ -0,0 +1,35 @@
...
+    }
+
+    pwm.initialize();
+    return 1;
```
Comment (staroselskii) [inline C++/Navio/Navio+/RCOutput_Navio.cpp]: return true or maybe return pwm.inititalize() ?

File: C++/Navio/Navio+/RCOutput_Navio.cpp [hunk 17]
```diff
@@ -0,0 +1,35 @@
...
+
+void RCOutput_Navio::set_frequency(int channel, float frequency)
+{
+    pwm.setFrequency(frequency);
+}
+
+bool RCOutput_Navio::set_duty_cycle(int channel, float period)
+{
+    pwm.setPWMmS(channel+3, period / 1000); // 1st Navio RC output is 3 and period must be less 10
+
```
Comment (staroselskii) [inline C++/Navio/Navio+/RCOutput_Navio.cpp]: where's the ruturn?

File: C++/Navio/Navio+/RCOutput_Navio.h [hunk 18]
```diff
@@ -0,0 +1,23 @@
...
+#include <Common/gpio.h>
+
+using namespace Navio;
+
+
+class RCOutput_Navio : public RCOutput
+{
+public:
+    RCOutput_Navio();
+    bool initialize(int channel);
```
Comment (staroselskii) [inline C++/Navio/Navio+/RCOutput_Navio.h]: `override` missing everywhere.

File: C++/Navio/Navio+/ADC_Navio.cpp [hunk 19]
```diff
@@ -0,0 +1,27 @@
...
+{
+    ADS1115 adc;
+
+    adc.setMode(ADS1115_MODE_SINGLESHOT);
+    adc.setRate(ADS1115_RATE_860);
+}
+
+int ADC_Navio::read(int ch)
+{
+    adc.setMultiplexer(muxes[ch]);
```
Comment (staroselskii) [inline C++/Navio/Navio+/ADC_Navio.cpp]: read(100) will blow up stack. Please add a check `if ch < ARRAY_SIZE(muxes)`.

**Review by staroselskii — CHANGES_REQUESTED** [review]
Nice but please verify that all methods that should have return values, do actually have them. Add parameter checking if need be. Like if we channel.read(100), it should just return an error.
Next step would be to turn all Navio/Navio2 and Common to a library libnavio and then just link examples to it like this gcc -o adc adc.cpp -lnavio

**Comment by IgorAnohin** [comment]
Hi!
I fixed bugs and add new library.

File: C++/Navio/Navio2/Led_Navio2.cpp [hunk 20]
```diff
@@ -0,0 +1,18 @@
+#include "Led_Navio2.h"
+
+Led_Navio2::Led_Navio2()
+{
+
+}
+
+bool Led_Navio2::initialize()
+{
+    return (!pwm.initialize());
```
Comment (staroselskii) [inline C++/Navio/Navio2/Led_Navio2.cpp]: We don't use PWM for Leds on Navio2.
It's also extremely confusing to negate the initialize()'s return value here. Why?

**Review by staroselskii — CHANGES_REQUESTED** [review]
Everything's fine except LEDs examples don't work on either of boards.

File: C++/Navio/Navio2/RGBled.cpp [hunk 21]
```diff
@@ -14,7 +14,7 @@ bool RGBled::initialize() {
     pinR = new Pin(RPI_GPIO_4);
     pinG = new Pin(RPI_GPIO_27);
     pinB = new Pin(RPI_GPIO_6);
-    if (pinR->init() & pinG->init() & pinB->init()) {
+    if (pinR->init() && pinG->init() && pinB->init()) {
```
Comment (staroselskii) [inline C++/Navio/Navio2/RGBled.cpp:17]: The change is correct but the commit message is not good.
`C++: Navio2: RGBLed: fail on logic AND in pin initialization`

The body should say that there had been bitwise and in the past.

File: C++/Navio/Navio2/RGBled.cpp [hunk 22]
```diff
@@ -25,7 +25,7 @@ bool RGBled::initialize() {
         pinB->write(OFF);
     }
     else {
-        fprintf(stderr, "Output enable not set. Are you root?\n");
+        fprintf(stderr, "Initialization failed\n");
```
Comment (staroselskii) [inline C++/Navio/Navio2/RGBled.cpp]: It's better to give more context to an error. Like `%s: pins' initialization failed, __FUNCTION__`

**Review by staroselskii — CHANGES_REQUESTED** [review]

**Review by staroselskii — APPROVED** [review]

**Pull request merged**